WIP: mmapstorage

mpranj opened the pull request:
# Purpose

Quick preview of the `mmapstorage` plugin. Not yet ready for merge or review.
A simple storage plugin benchmark is added to compare it to `dump`.  It does not yet store the OPMPHM data structure.

To enable it with cmake one can use `-DKDB_DEFAULT_STORAGE=mmapstorage -DKDB_DB_FILE='default.mmap' -DKDB_DB_INIT='elektra.mmap'`.

# Checklist

- [ ] commit messages are fine ("module: short statement" syntax and refer to issues)
- [ ] I added unit tests
- [x] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)
- [ ] release notes are updated (doc/news/_preparation_next_release.md)

mpranj commented:
@ingwinlu could you glance over 67ceb698f18c93f62ae43ee7aa97619603b90dda and see whether that would be the proper way to add a new build job with a different default storage plugin. I basically copied it from the `ini` job, but just want to make sure that is how you would do it.

ingwinlu commented:
@mpranj lgtm

mpranj commented:
Thanks, and thanks for putting all the hard work into the build system!

mpranj commented:
@markus2330 other than the README, no major rewrites of the plugin are planned. The benchmarks will be improved, and a few tests might be added.

I plan to review everything again myself before I mark it as ready. If you want to do a preliminary review of the plugin code, it would make sense for me now.

_NB: The checks haven't completed, but if something is broken it is nothing of major concern. Locally everything runs fine._

markus2330 commented:
It is better if you finish it first! Without a README.md reviewing is difficult.

mpranj commented:
@markus2330 please review my pull request.

The remaining travis check is clang-asan, which likes to time out. Unfortunately, it is a matter of luck whether it succeeds or not. For what it's worth: it succeeds on my fork, so I think it can be ignored.

markus2330 commented on src/libs/elektra/keyvalue.c:
```diff
@@ -526,6 +528,18 @@ ssize_t keySetRaw (Key * key, const void * newBinary, size_t dataSize)
+		}
+		else
+		{
+			if (-1 == elektraRealloc ((void **) &key->data.v, key->dataSize)) return -1;
```
good to have the check here, but it is now inconsistent with the other places where the return value of elektraRealloc is not checked? (Can you add the check everywhere?)

mpranj replied:
I added the check on every elektraRealloc where I already touched the source. I think adding the check on **every** realloc or malloc would be better in a separate issue/PR?

markus2330 replied:
No, if it is consistent for every realloc and malloc you touched it is perfect for this PR.

markus2330 commented on src/plugins/README.md:
```diff
@@ -118,6 +118,7 @@ productive use:
 - [camel](camel/) reads and writes a very limited subset of [YAML][]
 - [yamlcpp](yamlcpp/) reads and writes data in the [YAML][] format using [yaml-cpp](https://github.com/jbeder/yaml-cpp)
 - [yanlr](yanlr/) reads data using a parser generated by [ANTLR](http://www.antlr.org)
+- [mmapstorage](mmapstorage/) uses binary, not portable memory mapped file for high performance storage
```
for a

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+include (LibAddMacros)
+
+if (DEPENDENCY_PHASE)
+	set (plugin mmapstorage)
```
only used once, does not seem to be useful?

mpranj replied:
Removed.

markus2330 commented on src/plugins/mmapstorage/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+		    ADD_TEST
+		    TEST_README)
+else (ENABLE_CHECKSUM)
+	add_plugin (mmapstorage SOURCES ${MMAPSTORAGE_SOURCES} LINK_ELEKTRA elektra-core ADD_TEST TEST_README)
```
as written very far above: this plugin can always be included (as compilation variant)

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+set (MMAPSTORAGE_SOURCES dynarray.h dynarray.c mmapstorage.h mmapstorage.c)
+
+if (ENABLE_CHECKSUM)
+	add_plugin (mmapstorage
```
rename to mmapstorage_zlib? (compilation variant)

mpranj replied:
It is called mmapstorage_crc.

markus2330 commented on src/plugins/mmapstorage/README.md:
```diff
@@ -0,0 +1,81 @@
+- infos/provides = storage/mmapstorage
+- infos/recommends =
+- infos/placements = getstorage setstorage
+- infos/status = maintained unittest concept preview discouraged
```
discouraged?

mpranj replied:
Until the global cache is finished, I can not encourage the use. If I am forced to change something in the format, it will be incompatible with future versions.

If you think that discouraged is misplaced here, I can of course remove it.

NB: I built in format versioning, but just to avoid crashes, not for backwards compatibility.

markus2330 replied:
discouraged is the most negative word we have. It is your decision, if you think nobody should use it at the moment, it is okay.

Actually, if you do not say "compatible" you do not give any guarantees at all. So actually nothing else needs to be said.

mpranj replied:
I removed the discouraged, it is not so bad. :smile:

markus2330 commented on src/plugins/mmapstorage/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Introduction
+
+This is a high performance storage plugin that supports full Elektra semantics.
```
back up the claim :smile:

mpranj replied:
Added benchmark numbers to news.

markus2330 commented on src/plugins/mmapstorage/dynarray.c:
```diff
@@ -0,0 +1,159 @@
+
+	while (l < h)
+	{
+		m = (l + h) >> 1;
```
`size_t m` here?

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/dynarray.c:
```diff
@@ -0,0 +1,159 @@
+
+#include <kdblogger.h>
+
+/* -- DynArray Implementation ----------------------------------------------------------------------------------------------------------- */
```
Please add a short description of what the dynarray does.

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/dynarray.h:
```diff
@@ -0,0 +1,32 @@
+
+#include <kdbprivate.h>
+
+#include <sys/types.h> // size_t
```
Isn't this pulled in by kdbprivate.h?

mpranj replied:
Yes. Removed it.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+#include "kdbconfig.h"
+#endif
+
+#include <assert.h>
```
should not be needed (kdbassert.h instead)

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+#define SIZEOF_MMAPHEADER (sizeof (MmapHeader))
+#define SIZEOF_MMAPMETADATA (sizeof (MmapMetaData))
+#define SIZEOF_MMAPFOOTER (sizeof (MmapFooter))
+#define SIZEOF_ADDR_STRING (19) // format: 0xADDR -> ADDR in hex, for 64bit addr: 2 bytes (0x) + 16 bytes (ADDR) + 1 byte (ending null)
```
Use doxygen-style comments.

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+#define SIZEOF_MMAPFOOTER (sizeof (MmapFooter))
+#define SIZEOF_ADDR_STRING (19) // format: 0xADDR -> ADDR in hex, for 64bit addr: 2 bytes (0x) + 16 bytes (ADDR) + 1 byte (ending null)
+
+struct _mmapAddr
```
no docu.

mpranj replied:
Added docu.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+
+	if ((fp = fopen (keyString (parentKey), mode)) == 0)
+	{
+		ELEKTRA_LOG_WARNING ("error opening file %s", keyString (parentKey));
```
warning or error?

mpranj replied:
The plugin returns an error (and sets it in the parentKey), should this happen. The log level is warning because the level error is reserved for `ELEKTRA_ASSERT` according to doc/dev/logging.md.
Maybe I am misunderstanding something here?

markus2330 replied:
No, everything is fine! Thank you for clearing this up!

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+	memset (mmapHeader, 0, SIZEOF_MMAPHEADER);
+	memset (mmapMetaData, 0, SIZEOF_MMAPMETADATA);
+
+	if (fread (mmapHeader, SIZEOF_MMAPHEADER, 1, fp) != 1 || fread (mmapMetaData, SIZEOF_MMAPMETADATA, 1, fp) != 1)
```
Why fread if we plan to or already have it mmaped?

mpranj replied:
Done. Removed all fread() and using only mmap() now.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+	magicKeySet.flags = KS_FLAG_MMAP_ARRAY | KS_FLAG_SYNC;
+	magicKeySet.mmapMetaData = (MmapMetaData *) 0xE1EF;
+#ifdef ELEKTRA_ENABLE_OPTIMIZATIONS
+	magicKeySet.opmphm = (Opmphm *) 0xFFFE;
```
avoid magic constants like this (should be const int or macro)

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+ *
+ * @param ks keyset to verify
+ *
+ * @return 0 if magic KeySet is correct
```
retval

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+ *
+ * @param key to verify
+ *
+ * @return 0 if magic Key is correct
```
retval

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+	}
+
+	// write the meta keys into place
+	Key * mmapMetaKey;
```
move to where needed?

mpranj replied:
Done.

markus2330 commented on src/plugins/mmapstorage/mmapstorage.c:
```diff
@@ -0,0 +1,1323 @@
+			keySet->array = copy->array;
+			keySet->size = copy->size;
+			keySet->alloc = copy->alloc;
+			keySet->cursor = 0;
```
Shouldn't that be that this should be done ksClose? (It seems like currently it is not done and the docu of ksClose is not very enlightening, too :cry:)

mpranj replied:
You're right. ksClose calls ksRewind which does exactly that. I removed my unnecessary manual rewinds.

mpranj commented on cmake/ElektraCache.cmake:
```diff
@@ -245,6 +245,11 @@ option (INSTALL_BUILD_TOOLS "Install build tools for cross-compilation" OFF)
 
 option (ENABLE_OPTIMIZATIONS "Turn on optimizations that trade memory for speed" ON)
 
+option (ENABLE_CHECKSUM "Turn on checksum for mmapstorage." OFF)
```
The cmake option is temporary and will be removed. In the mean time (for benchmarks) I will use the compilation variants, and remove the cmake option.

> If it is costly, a run-time option to the mmap plugin is preferable.

By costly you mean the CRC function?

Ultimately, there should be only one variant. Whether or not to use checksums should could be a team decision, if I present performance numbers?

mpranj replied:
For now, we have compilation variants. Adding a CRC plugin should be easy but opens up the question where the checksum is stored.

mpranj commented on src/include/kdbmmap.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * @file
+ *
+ * @brief Declarations for the mmap storage and cache
```
We don't need a public include. Putting everything in kdbprivate.h works as well (I just did it).
I just did not want to clutter kdbprivate.h with it.

mpranj replied:
Upon review, I only need the MmapMetaData there. I moved the MmapHeader and MmapFooter definitions completely into the plugin.

mpranj replied:
Done.

mpranj commented on src/include/kdbmmap.h:
```diff
@@ -0,0 +1,76 @@
+#define ELEKTRA_MAGIC_MMAP_NUMBER (0x0A6172746B656C45)
+#define ELEKTRA_MMAP_FORMAT_VERSION (1)
+
+typedef struct _mmapHeader MmapHeader;
```
I do not follow. Is this even relevant if everything is in kdbprivate.h?

markus2330 replied:
No it is not relevant then.

mpranj commented on src/include/kdbmmap.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Mmap meta-data
+ */
+struct _mmapMetaData
```
In the format it is located between header and footer, yes.

mpranj commented on src/libs/elektra/keyname.c:
```diff
@@ -854,7 +856,17 @@ ssize_t keyAddBaseName (Key * key, const char * baseName)
 	}
 
-	elektraRealloc ((void **) &key->key, key->keySize * 2);
+	if (test_bit (key->flags, KEY_FLAG_MMAP_KEY) == KEY_FLAG_MMAP_KEY)
```
**No**. :upside_down_face: I removed the other occurrences too.

mpranj commented on src/libs/elektra/keyname.c:
```diff
@@ -854,7 +856,17 @@ ssize_t keyAddBaseName (Key * key, const char * baseName)
+	if (test_bit (key->flags, KEY_FLAG_MMAP_KEY) == KEY_FLAG_MMAP_KEY)
+	{
+		// key was in mmap region, clear flag and trigger malloc instead of realloc
+		key->key = elektraMalloc (key->keySize * 2);
```
Removing *2 results in invalid reads.

markus2330 replied:
I meant that it should be moved out of the branches to a single place, not to remove it all :smile:

mpranj commented on src/libs/elektra/keyname.c:
```diff
@@ -962,7 +974,18 @@ ssize_t keyAddName (Key * key, const char * newName)
+	if (test_bit (key->flags, KEY_FLAG_MMAP_KEY) == KEY_FLAG_MMAP_KEY)
+	{
+		// key was in mmap region, clear flag and trigger malloc instead of realloc
+		key->key = elektraMalloc (newSize * 2);
```
Removing *2 results in invalid reads. (here too)

mpranj commented on doc/news/_preparation_next_release.md:
```diff
@@ -66,6 +66,11 @@ The following section lists news about the [modules](https://www.libelektra.org/
 
+### mmapstorage
+
+- Added a new, binary and fast storage plugin called [mmapstorage](https://libelektra.org/plugins/mmapstorage).
```
Done.

mpranj commented on doc/news/_preparation_next_release.md:
```diff
@@ -169,6 +174,8 @@ you up to date with the multi-language support provided by Elektra.
 - The tests `testkdb_allplugins` and `testscr_check_kdb_internal_check` do not test a plugin on an ASAN enabled build anymore, if you
   specify the status tag `memleak` in the plugin contract. *(René Schwaiger)*
+- Added many storage plugin tests. Most tests use the keyset, key name and value APIs.
+  Currently, the tests are only active for dump and mmapstorage. *(Mihael Pranjić)*
```
Done.

mpranj commented on src/include/kdbconfig.h.in:
```diff
@@ -105,6 +105,12 @@
 /* ASAN */
 #cmakedefine ENABLE_ASAN
 
+/* enable mmap checksum */
```
Done.

mpranj commented:
Thank you for the thorough review! I made the requested changes.
Additionally the following was done:

- added magic data / consistency checks for MmapMetaData (I overlooked this earlier)
- improved the output format (now CSV with header) of the storage benchmark for easier evaluation
- added the benchmark results to the news (but **not** into the plugin README). If you have a better idea how/where to put it, do tell.
- added myself to the AUTHORS file

If you have further change requests, please tell. Otherwise it can be merged.
